compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEControlCodeToTransaction.cpp

```diff
@@ -0,0 +1,360 @@
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult controlCodeToTransaction(IRRewriter &rewriter,
+                                       AMDAIE::ControlCodeOp controlCodeOp,
+                                       TransactionBuilder &builder) {
+  SmallVector<Operation *> toBeErased;
+  WalkResult res = controlCodeOp->walk([&](Operation *op) {
+    LogicalResult switchResult =
+        TypeSwitch<Operation *, LogicalResult>(op)
+            .Case<AMDAIE::NpuAddressPatchOp, AMDAIE::NpuDmaWaitOp,
+                  AMDAIE::NpuPushToQueueOp, AMDAIE::NpuWriteBdOp>(
+                [&](auto npuOp) {
+                  if (failed(convertOp(npuOp, builder))) return failure();
+                  toBeErased.push_back(npuOp);
+                  return success();
+                })
+            .Default([&](Operation *) { return success(); });
+    if (failed(switchResult)) return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+  if (res.wasInterrupted()) return failure();
+  for (Operation *op : toBeErased) {
+    op->dropAllUses();
+    rewriter.eraseOp(op);
+  }
+  return success();
+}
+
+namespace {
+
+class AMDAIEControlCodeToTransactionPass
+    : public impl::AMDAIEControlCodeToTransactionBase<
+          AMDAIEControlCodeToTransactionPass> {
+ public:
+  AMDAIEControlCodeToTransactionPass(
+      const AMDAIEControlCodeToTransactionOptions &options)
+      : AMDAIEControlCodeToTransactionBase(options) {}
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+  void runOnOperation() override;
+};
+
+void AMDAIEControlCodeToTransactionPass::runOnOperation() {
+  Operation *parentOp = getOperation();
+  MLIRContext *context = &getContext();
+
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(parentOp);
+  std::optional<AMDAIEDevice> maybeDevice = getConfigAMDAIEDevice(targetAttr);
+  if (!maybeDevice) {
+    parentOp->emitOpError()
+        << "has no AMDAIEDevice in the target attribute configuration. This "
+           "device-specific information is required to lower control code "
+           "ops.";
+    return signalPassFailure();
+  }
+  AMDAIE::AMDAIEDeviceModel deviceModel =
+      AMDAIE::getDeviceModel(maybeDevice.value());
+  TransactionBuilder transactionBuilder(std::move(deviceModel));
+
+  IRRewriter rewriter(context);
+  WalkResult res = parentOp->walk([&](AMDAIE::WorkgroupOp workgroupOp) {
+    transactionBuilder.clearAndInitialize();
+    if (failed(controlCodeToTransaction(rewriter, workgroupOp.getControlCode(),
+                                        transactionBuilder))) {
+      return WalkResult::interrupt();
+    }
+    LLVM_DEBUG(llvm::dbgs() << "Instruction size: "
+                            << transactionBuilder.getInstructionSize() << "\n");
+    ArrayRef<uint32_t> instructions =
+        transactionBuilder.finalizeAndReturnInstructions();
+    workgroupOp->setAttr(
+        "npu_instructions",
```

Comment:
can you do me a favor - can you make this a real attr on that op? so that we're not doing string look-ups/matches/etc.

Reply:
Jup, done